include/raft.h.in

```diff
@@ -1022,6 +1022,153 @@ RAFT_API void raft_seed(struct raft *r, unsigned random);
  * The @update output parameter contains changes that user code should perform
  * after the raft_step() call returns (e.g. new entries that must be persisted,
  * new messages that must be sent, etc.).
+ *
+ * Memory Ownership Model for raft_event Fields
+ * =============================================
+ *
+ * The caller must understand the memory ownership semantics for each event type
+ * to avoid memory leaks or use-after-free bugs. This section documents who owns
+ * the memory for each field in the raft_event union after raft_step() returns.
+ *
+ * RAFT_START Event
+ * ----------------
+ * - event->start.term: Scalar value (no ownership concerns)
+ * - event->start.voted_for: Scalar value (no ownership concerns)
+ * - event->start.metadata: PARTIAL OWNERSHIP TRANSFER. The nested configuration
+ *   field is transferred to raft (moved into r->configuration). The metadata
+ *   structure itself is NOT freed by raft and must be freed by the caller after
+ *   raft_step() returns, regardless of success or error. The caller must NOT
+ *   close the configuration field after the call, as it has been transferred.
```

Comment:
This is correct. However, I think it would probably be better to make a copy of the nested `configuration` field, so the ownership of the `metadata` structure stays entirely in the hands of the caller (see the other comment below).

Reply:
This an example of 7 entries in 2 batches:

```C
entries[0].batch = <memory-location-1>; /* This allocation is for the first 4 entries */
entries[1].batch = <memory-location-1>;
entries[2].batch = <memory-location-1>;
entries[3].batch = <memory-location-1>;
entries[4].batch = <memory-location-2>; /* This allocation is for the last 3 entries */
entries[5].batch = <memory-location-2>;
entries[6].batch = <memory-location-2>;
```

Note that the `batch` pointer should not be `NULL` in subsequent entries, but it's rather set to common batch location. However, I believe that in if you set it it to `NULL` things will still work, because of the way the code is currently written, but it's kind of accidental.

Reply:
Ah so my understanding was wrong: we can have multiple batch pointers, and they can repeat, simply we cannot have something like <memory-location-1> <memory-location-2> <memory-location-1>?